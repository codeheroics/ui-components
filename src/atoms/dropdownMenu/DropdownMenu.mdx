---
name: Dropdown Menu
menu: Components
---

import { Playground, Props } from 'docz';
import DropdownMenu from './DropdownMenu';
import Button from '../button/Button';
import PhoneIcon from '../icons/PhoneIcon';
import QuickActionsMenu from '../../molecules/quick-actions-menu/QuickActionsMenu';

# Dropdown Menu

<Playground
  style={{
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-around',
    height: '200px'
  }}
>
  <DropdownMenu
    elements={[
      {
        title: 'Hello',
        icon: <PhoneIcon />,
        onClick: () => {}
      },
      {
        title: 'Holà',
        onClick: () => {}
      },
      {
        title: 'Bonjour',
        onClick: () => {}
      },
      {
        title: 'Gutten Tag',
        onClick: () => {}
      }
    ]}
  >
    <Button nature='primary'> Click Me </Button>

  </DropdownMenu>

<DropdownMenu
  menuMaxHeight={150}
  elements={[
    {
      title: 'Hello',
      onClick: () => {}
    },
    {
      title: 'Holà',
      onClick: () => {}
    },
    {
      title: 'Bonjour',
      onClick: () => {}
    },
    {
      title: 'Gutten Tag',
      onClick: () => {}
    },
    {
      title: 'Ciao',
      onClick: () => {}
    },
    {
      title: 'Konnichiha',
      onClick: () => {}
    },
    {
      title: 'Ni Hao',
      onClick: () => {}
    },
    {
      title: 'Dobry Degn',
      onClick: () => {}
    }
  ]}
>
  <Button nature='primary'> With long list </Button>
</DropdownMenu>

  <DropdownMenu
    elements={[
      {
        title: 'Hello',
        onClick: () => {}
      },
      {
        title: 'Holà',
        onClick: () => {}
      }
    ]}
    dockingSide="right"
  >
    <Button nature='primary'> On the right </Button>

  </DropdownMenu>

</Playground>
