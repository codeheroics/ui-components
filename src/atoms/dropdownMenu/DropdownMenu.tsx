import * as React from 'react';
import DropdownCard from '@ions/dropdown/card/DropdownCard';
import styled from 'styled-components';
import { DockingSide, Placement } from './types';
import {
  getControllerAndMenuDimensions,
  getDropdownPosition,
  isDropdownOutOfContainer
} from './helpers';
import {
  getContainerBoundaries,
  getElementPlacement
} from '@utility/positionCompute';

interface Props {
  isOpen: boolean;
  controllerRef: React.RefObject<HTMLElement>;
  menuMaxHeight?: number;
  dockingSide?: DockingSide;
  placement?: Placement;
  containerRef?: React.RefObject<HTMLElement>;
}
export interface MenuWrapperProps {
  placement: Placement;
  dockingSide: DockingSide;
  isReadyForDisplay: boolean;
}

const MenuWrapper = styled.div<MenuWrapperProps>`
  position: absolute;
  opacity: ${props => (props.isReadyForDisplay ? 1 : 0)};
  ${getDropdownPosition}
  max-width: 318px;
`;

const DEFAULT_PLACEMENT = 'bottom';
const POSSIBLE_PLACEMENTS: Placement[] = ['top', 'bottom'];

const DropdownMenu: React.FC<Props> = props => {
  const [placement, setPlacement] = React.useState<Placement | null>(null);
  const menuRef = React.useRef<HTMLDivElement>(null);
  const prevIsOpen = React.useRef(false);

  React.useEffect(() => {
    if (props.isOpen === prevIsOpen.current) {
      return;
    }
    prevIsOpen.current = props.isOpen;
    setPlacement(null);

    if (!menuRef.current || !props.controllerRef.current) {
      return;
    }
    const elementDimensions = getControllerAndMenuDimensions(
      menuRef,
      props.controllerRef
    );

    const containerElement =
      (props.containerRef && props.containerRef.current) || window;
    const containerDimensions = getContainerBoundaries(containerElement);

    try {
      const newPlacement = getElementPlacement(
        isDropdownOutOfContainer,
        props.placement || DEFAULT_PLACEMENT,
        POSSIBLE_PLACEMENTS,
        elementDimensions,
        containerDimensions
      );
      setPlacement(newPlacement);
    } catch (e) {
      console.error(e);
    }
  });

  return (
    <>
      {props.isOpen && (
        <MenuWrapper
          placement={placement || DEFAULT_PLACEMENT}
          ref={menuRef}
          dockingSide={props.dockingSide || 'left'}
          isReadyForDisplay={placement !== null}
        >
          <DropdownCard maxHeight={props.menuMaxHeight}>
            {props.children}
          </DropdownCard>
        </MenuWrapper>
      )}
    </>
  );
};

export default DropdownMenu;
