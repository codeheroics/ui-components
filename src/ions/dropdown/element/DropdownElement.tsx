import * as React from 'react';
import { StyledText } from '@ions/text/StyledText';
import styled from 'styled-components';
import { dashGreen00, dashGreen06 } from '@colors';

interface Props {
  text: string;
  icon?: JSX.Element;
  onClick: (e: React.MouseEvent) => void;
}

const Text = styled(StyledText)`
  color: ${dashGreen00};
  line-height: 20px;
  font-size: 16px;
`;

const Wrapper = styled.button`
  cursor: pointer;
  border: transparent 0px;
  outline: none;
  display: flex;
  box-sizing: border-box;
  height: 36px;
  flex-shrink: 0;
  padding: 8px 16px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;

  &:hover {
    background-color: ${dashGreen06};
  }
`;

const IconContainer = styled.div`
  fill: ${dashGreen00};
  margin-right: 8px;
`;

const DropdownElement: React.FC<Props> = props => {
  return (
    <Wrapper type='button' onClick={props.onClick}>
      {props.icon && <IconContainer>{props.icon}</IconContainer>}
      <Text>{props.text}</Text>
    </Wrapper>
  );
};

export default DropdownElement;
