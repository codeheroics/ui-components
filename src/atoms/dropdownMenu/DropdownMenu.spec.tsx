import * as React from 'react';
import { mount } from 'enzyme';
import DropdownMenu from './DropdownMenu';
import DropdownElement from '@ions/dropdown/element/DropdownElement';

describe('<Button>', () => {
  describe('Global render', () => {
    it('should render', () => {
      const controllerRef = React.createRef<HTMLDivElement>();

      const DropdownMenuWrapper = mount(
        <div>
          <div ref={controllerRef}>Dropdwon Menu controller</div>
          <DropdownMenu controllerRef={controllerRef} isOpen={true}>
            <DropdownElement onClick={() => {}}>Element 1</DropdownElement>
            <DropdownElement onClick={() => {}}>Element 2</DropdownElement>
          </DropdownMenu>
        </div>
      );

      expect(DropdownMenuWrapper).toMatchSnapshot();
    });
  });
});
