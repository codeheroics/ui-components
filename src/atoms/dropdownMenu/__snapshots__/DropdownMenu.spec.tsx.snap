// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> Global render should render 1`] = `
<div>
  <div>
    Dropdwon Menu controller
  </div>
  <DropdownMenu
    controllerRef={
      Object {
        "current": <div>
          Dropdwon Menu controller
        </div>,
      }
    }
    isOpen={true}
  >
    <styled.div
      dockingSide="left"
      isReadyForDisplay={true}
      placement="bottom"
    >
      <StyledComponent
        dockingSide="left"
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bwzfXH",
              "isStatic": false,
              "lastClassName": "izZuan",
              "rules": Array [
                "
  position: absolute;
  opacity: ",
                [Function],
                ";
  ",
                [Function],
                "
  max-width: 318px;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bwzfXH",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={
          Object {
            "current": <div
              class="sc-bwzfXH izZuan"
            >
              <div
                class="sc-bdVaJa kurQmA"
              >
                <button
                  class="sc-ifAKCX epOyKj"
                  type="button"
                >
                  <span
                    class="sc-htpNat sc-bxivhb hojljZ"
                  >
                    Element 1
                  </span>
                </button>
                <button
                  class="sc-ifAKCX epOyKj"
                  type="button"
                >
                  <span
                    class="sc-htpNat sc-bxivhb hojljZ"
                  >
                    Element 2
                  </span>
                </button>
              </div>
            </div>,
          }
        }
        isReadyForDisplay={true}
        placement="bottom"
      >
        <div
          className="sc-bwzfXH izZuan"
        >
          <styled.div>
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bdVaJa",
                    "isStatic": false,
                    "lastClassName": "kurQmA",
                    "rules": Array [
                      "
  display: flex;
  flex-direction: column;
  background-color: ",
                      "#FFFFFF",
                      ";
  border: 1px solid ",
                      "#CFD7D8",
                      ";
  box-shadow: 0px 4px 16px rgba(0, 0, 0, 0.2);
  border-radius: 4px;
  padding: 8px 0;
  max-height: ",
                      [Function],
                      "px;
  overflow-y: scroll;
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bdVaJa",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
            >
              <div
                className="sc-bdVaJa kurQmA"
              >
                <DropdownElement
                  onClick={[Function]}
                >
                  <styled.button
                    onClick={[Function]}
                    type="button"
                  >
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-ifAKCX",
                            "isStatic": true,
                            "lastClassName": "epOyKj",
                            "rules": Array [
                              "
  cursor: pointer;
  border: transparent 0px;
  outline: none;
  display: flex;
  box-sizing: border-box;
  height: 36px;
  flex-shrink: 0;
  padding: 8px 16px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;

  &:hover {
    background-color: ",
                              "#F5F7F7",
                              ";
  }
",
                            ],
                          },
                          "displayName": "styled.button",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-ifAKCX",
                          "target": "button",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="sc-ifAKCX epOyKj"
                        onClick={[Function]}
                        type="button"
                      >
                        <Styled(styled.span)>
                          <StyledComponent
                            forwardedComponent={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": Array [],
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-bxivhb",
                                  "isStatic": true,
                                  "lastClassName": "hojljZ",
                                  "rules": Array [
                                    "
  font-family: inherit;
  font-weight: 600;
",
                                    "
  color: ",
                                    "#0E353D",
                                    ";
  line-height: 20px;
  font-size: 16px;
",
                                  ],
                                },
                                "displayName": "Styled(styled.span)",
                                "foldedComponentIds": Array [
                                  "sc-htpNat",
                                ],
                                "render": [Function],
                                "styledComponentId": "sc-bxivhb",
                                "target": "span",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={null}
                          >
                            <span
                              className="sc-htpNat sc-bxivhb hojljZ"
                            >
                              Element 1
                            </span>
                          </StyledComponent>
                        </Styled(styled.span)>
                      </button>
                    </StyledComponent>
                  </styled.button>
                </DropdownElement>
                <DropdownElement
                  onClick={[Function]}
                >
                  <styled.button
                    onClick={[Function]}
                    type="button"
                  >
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-ifAKCX",
                            "isStatic": true,
                            "lastClassName": "epOyKj",
                            "rules": Array [
                              "
  cursor: pointer;
  border: transparent 0px;
  outline: none;
  display: flex;
  box-sizing: border-box;
  height: 36px;
  flex-shrink: 0;
  padding: 8px 16px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;

  &:hover {
    background-color: ",
                              "#F5F7F7",
                              ";
  }
",
                            ],
                          },
                          "displayName": "styled.button",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-ifAKCX",
                          "target": "button",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="sc-ifAKCX epOyKj"
                        onClick={[Function]}
                        type="button"
                      >
                        <Styled(styled.span)>
                          <StyledComponent
                            forwardedComponent={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": Array [],
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-bxivhb",
                                  "isStatic": true,
                                  "lastClassName": "hojljZ",
                                  "rules": Array [
                                    "
  font-family: inherit;
  font-weight: 600;
",
                                    "
  color: ",
                                    "#0E353D",
                                    ";
  line-height: 20px;
  font-size: 16px;
",
                                  ],
                                },
                                "displayName": "Styled(styled.span)",
                                "foldedComponentIds": Array [
                                  "sc-htpNat",
                                ],
                                "render": [Function],
                                "styledComponentId": "sc-bxivhb",
                                "target": "span",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={null}
                          >
                            <span
                              className="sc-htpNat sc-bxivhb hojljZ"
                            >
                              Element 2
                            </span>
                          </StyledComponent>
                        </Styled(styled.span)>
                      </button>
                    </StyledComponent>
                  </styled.button>
                </DropdownElement>
              </div>
            </StyledComponent>
          </styled.div>
        </div>
      </StyledComponent>
    </styled.div>
  </DropdownMenu>
</div>
`;
